packages/xo-server/src/xo-mixins/authentication.mjs

```diff
@@ -167,14 +167,23 @@ export default class {
         throw new Error('too fast authentication tries')
       }
 
-      const result = await this._authenticateUser(credentials, userData)
-      if (result === undefined) {
-        failures[username] = now
-        throw invalidCredentials()
+      const { otp, ...rest } = credentials
+      const result = await this._authenticateUser(rest, userData)
+      if (result !== undefined) {
+        const secret = result.user.preferences?.otp
+        if (
+          secret === undefined ||
+          bypassOtp ||
+          result.bypassOtp || // some authentication providers bypass OTP (e.g. token)
+          (await verifyTotp(otp, { secret }))
+        ) {
+          delete failures[username]
+          return result
+        }
       }
 
-      delete failures[username]
-      return result
+      failures[username] = now
+      throw invalidCredentials()
```

Comment:
So in case the `OTP` code is invalid, we throw an `invalidCredentials` error. Is that wanted?

Reply:
Yes, OTP is part of the credentials and those are invalid.
I believe it is good enough for an API implementation.

Any opinions on this?